src/main/java/io/cdap/plugin/servicenow/source/ServiceNowRecordReader.java

```diff
@@ -0,0 +1,268 @@
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+import static io.cdap.plugin.servicenow.source.util.ServiceNowConstants.PAGE_SIZE;
+
+/**
+ * Record reader that reads the entire contents of a ServiceNow table.
+ */
+public class ServiceNowRecordReader extends RecordReader<NullWritable, StructuredRecord> {
+  private static final Logger LOG = LoggerFactory.getLogger(ServiceNowRecordReader.class);
+  private final ServiceNowSourceConfig pluginConf;
+  private ServiceNowInputSplit split;
+  private int pos;
+  private List<Schema.Field> tableFields;
+  private Schema schema;
+
+  private String tableName;
+  private String tableNameField;
+  private List<Map<String, Object>> results;
+  private Iterator<Map<String, Object>> iterator;
+  private Map<String, Object> row;
+
+  ServiceNowRecordReader(ServiceNowSourceConfig pluginConf) {
+    this.pluginConf = pluginConf;
+  }
+
+  @Override
+  public void initialize(InputSplit split, TaskAttemptContext context) {
+    this.split = (ServiceNowInputSplit) split;
+    this.pos = 0;
+  }
+
+  @Override
+  public boolean nextKeyValue() throws IOException {
+    try {
+      if (results == null) {
+        fetchData();
+      }
+
+      if (!iterator.hasNext()) {
+        return false;
+      }
+
+      row = iterator.next();
+
+      pos++;
+    } catch (Exception e) {
+      LOG.error("Error in nextKeyValue", e);
+      throw new IOException("Exception in nextKeyValue", e);
+    }
+    return true;
+  }
+
+  @Override
+  public NullWritable getCurrentKey() {
+    return NullWritable.get();
+  }
+
+  @Override
+  public StructuredRecord getCurrentValue() throws IOException {
+    StructuredRecord.Builder recordBuilder = StructuredRecord.builder(schema);
+
+    if (pluginConf.getQueryMode() == SourceQueryMode.REPORTING) {
+      recordBuilder.set(tableNameField, tableName);
+    }
+
+    try {
+      for (Schema.Field field : tableFields) {
+        String fieldName = field.getName();
+        Object fieldValue = convertToValue(fieldName, field.getSchema(), row);
+        recordBuilder.set(fieldName, fieldValue);
+      }
+    } catch (Exception e) {
+      LOG.error("Error decoding row from table " + tableName, e);
+      throw new IOException("Error decoding row from table " + tableName, e);
+    }
+    return recordBuilder.build();
+  }
+
+  @Override
+  public float getProgress() throws IOException, InterruptedException {
+    return pos / (float) split.getLength();
+  }
+
+  @Override
+  public void close() throws IOException {
+    SQLException exception = null;
+
+    if (exception != null) {
+      throw new IOException(exception);
+    }
+  }
+
+  private void fetchData() {
+    tableName = split.getTableName();
+    tableNameField = pluginConf.getTableNameField();
+
+    ServiceNowTableAPIClientImpl restApi = new ServiceNowTableAPIClientImpl(pluginConf);
+
+    //Get the table data
```

Comment:
Nit: space after //. Similar comment for other comments in this file.

Reply:
Done